Nullsafety

okocsis opened the pull request:
Hey, guys! I've made a lot of effort to port the dart codebase to a nullsafe environment. I would appreciate someone taking a look at it.

known issues:
-not all of the test are compiling, although I've tried to make it work I've even ported some of the mocks to the new generated mocking scheme. FIXME: I think many of the test should be retired, in favour of general nullsafety. Please help me on that.
-there was a bleadapter lifecycle fiasco, because the DevicesBloc destroyed the client on dispose(), making the internal adapter unreachable, after going to detail view. Although I have attempted to fix this by just removing the destroy client call from DevicesBloc.dispose() but I'm just not certain this was the right solution in terms of good design.

Thanks for taking the time!

Fixes #556

okocsis committed:
Work in progress

okocsis committed:
lib/** nullability compliant;
added linter exceptions for string single quotes and pedantic local variable type rule

okocsis committed:
Example:: updated to nullsafe-ready dependencies

okocsis committed:
Example:: nullsafety + refactor WIP

okocsis committed:
Example:: nullsafe ready

okocsis committed:
Example/test:: nullsafe ready

okocsis committed:
test:: using new mocking mechanism
additional fixes

okocsis committed:
Java: added some null checks to make the calls to stop and destroy reentrant

okocsis committed:
Example:: additional cleanup and refactor
Lib:: enchanced error handling, refined nullability on certain types

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Polidea/FlutterBleLib?pullRequest=586) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Polidea/FlutterBleLib?pullRequest=586) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: oliver<br/>:x: okocsis<br/><hr/>**oliver** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Polidea/FlutterBleLib?pullRequest=586) it.</sub>

okocsis committed:
TestFix:: the injected manager instance is not part of the Characteristic POD so should be ignored when checking for equallity

okocsis committed:
ble_error seem to have rather nullable fields

okocsis committed:
TestFix: little refactor on characteristic_mixin regarding clearity and type annotations

okocsis committed:
TestFix:: lib_core_test::  tests now passing

okocsis committed:
Example:: removed comment-disabled code

okocsis committed:
Test:: all tests Passing now! 🎉

okocsis committed:
Enhanced BleError nullability, so that reason always has a default value;
made fields final

okocsis committed:
Travis fix

okocsis committed:
disabled analysis for tests

okocsis committed:
Travis:: still fixing ios build

mikolak commented on .vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "editor.rulers": [ 
```
Add the file to .gitignore and remove it from PR, please :)

okocsis replied:
sure thing!

mikolak commented on .travis.yml:
```diff
@@ -55,6 +55,11 @@ _ios_job_template: &ios_job_template
     - export PATH=`pwd`/flutter/bin:`pwd`/flutter/bin/cache/dart-sdk/bin:$PATH
   script:
     - cd example
+    - flutter pub upgrade
```
Example shouldn't upgrade it's dependencies on CI only, this will quickly make it even more stale than it is now. ;)

okocsis replied:
gotcha, I was just playing around to make it pass somehow anyways ;)

okocsis replied:
By the way, do you happen to have any better idea why it fails on CI environment, 
I don't see a reason for it not to work, but it doesn't seem to use the correct updated version of permission_handler package, (upgraded to have more sound nullsafety, also bugfixes were taken place since the previous version)
I'm on Xcode 12.4, flutter 2.0.4.

mikolak replied:
Maybe CI is running outdated XCode?

okocsis replied:
can easily be that, shall i ammend it or will address it?

mikolak replied:
If you could, please. 🙂  This line: https://github.com/Polidea/FlutterBleLib/blob/develop/.travis.yml#L50

okocsis replied:
will do

mikolak commented on lib/ble_manager.dart:
```diff
@@ -147,7 +151,7 @@ abstract class BleManager {
   /// recognized on a different Android device depending on peripheral’s
   /// implementation and changes in future OS releases.
-  Peripheral createUnsafePeripheral(String peripheralId, {String name});
+  Peripheral createUnsafePeripheral(String peripheralId, {String? name});
```
Does an optional argument has to be explicitly typed as nullable? There's no point in explicitly passing a null to an optional argument if that's not the case.

okocsis replied:
yes, that's what the new dart standard says, doesn't even let you "compile" it without it. 
So to sum up name needs to eighter:
- have a default value {String name = ""} (I'm down with that too)
- or be explicitly nullable, so if the caller omits it. it will be null by default implicitly, that needs to be nullchecked thus the type must be marked nullable in that case

mikolak commented on lib/characteristic.dart:
```diff
@@ -20,7 +20,7 @@ class Characteristic extends InternalCharacteristic {
   /// The [Service] containing this characteristic.
   Service service;
```
This should also be final, like the manager

okocsis replied:
yes, indeed, in fact, a lot of vars and fields could be final, I only changed the ones I touched, so the PR remains relatively clean. But I generally agree with the "prefer final" idiom everywhere. I think it will be cleaner to do that refactor in another PR imho. :)

mikolak commented on lib/characteristic.dart:
```diff
@@ -20,7 +20,7 @@ class Characteristic extends InternalCharacteristic {
+  final ManagerForCharacteristic _manager;
 
   /// The UUID of this characteristic.
   String uuid;
```
Like above, also final 🙂

okocsis replied:
Do we want any of the fields of this object (POD) to be mutable? in other words, does the user ever want to alter any of the fields of a retrieved Characteristic object, if anything it only leads to inconsistency innit?

mikolak replied:
I wrote about it since you've finaled things on descriptor, but not on Characteristic. What needs to be done is actually move all the implementation away from characteristic to internal characteristic and make Characteristic abstract with everything final, so that it's only an interface, but it's somewhere on my to-do. Either way no worries, these are just nitpicks and I'll go over the all the classes before release, you can safely ignore it in this PR.

mikolak reviewed (commented):
Just some early feedback as I'm on 11/51 files - I'll finish it tomorrow. Looks nice so far, thanks for the PR!

okocsis committed:
vscode:: stopped tracking settings file

okocsis committed:
travis:: removed my additions to ios example script

okocsis committed:
Characteristic:: Service field is now final

okocsis committed:
gitignored andoid related staff
xcode noise

yauhenipakala committed:
Add bluetooth-central key to README.md (#562)

* Add bluetooth-central key to README.md

Based on #530

* Update README.md

mikolak commented on example/ios/Runner.xcodeproj/project.pbxproj:
```diff
@@ -228,13 +228,11 @@
 			);
 			inputPaths = (
 				"${PODS_ROOT}/Target Support Files/Pods-Runner/Pods-Runner-frameworks.sh",
-				"${PODS_ROOT}/../Flutter/Flutter.framework",
```
Are changes here intended?

okocsis replied:
I'm not even sure that this file is meant to be tracked on purpose or not.. maybe we should untrack this too?

mikolak replied:
I'm fairly certain it should be tracked, as it's the definition of project contents. I saw on my branch that integration must have changed and also got those changes automatically. Just wanted to make sure it wasn't by mistake. 🙂

mikolak commented on lib/characteristic.dart:
```diff
@@ -170,21 +168,22 @@ class Characteristic extends InternalCharacteristic {
 /// when it was first read from [Peripheral] or [Service].
-class CharacteristicWithValue extends Characteristic with WithValue {
+class CharacteristicWithValue extends Characteristic {
+  Uint8List value;
```
Why move value from the mixin?

okocsis replied:
I didn't get the point of the mixin here since it's not used anywhere else. 
Do you plan to use polymorphism over the type WithValue ever? I couldn't imagine any use case like for example if there was a function with the signature: `void foo(WithValue value)` I'd understand, one may pass in anything that conforms to WithValue thus has a `value` filed.

Otherwise, to me, it was cleaner and more concise to just have the field there, as jumping across files all the time takes a lot of time, for imho no benefit.
But again I might be missing some clue here.

mikolak replied:
Only benefit was that it was used for Characteristic and Descriptor. Not much of a value, I agree.

mikolak commented on lib/error/ble_error.dart:
```diff
@@ -13,23 +14,24 @@ abstract class _BleErrorMetadata {
+        iosErrorCode = json[_BleErrorMetadata.iosErrorCode],
         androidErrorCode = json[_BleErrorMetadata.androidErrorCode],
-        reason = json[_BleErrorMetadata.reason],
+        reason = json[_BleErrorMetadata.reason] ?? "",
```
```suggestion
        reason = json[_BleErrorMetadata.reason] ?? "Reason not provided",
```

okocsis replied:
hm, good suggestion! much better with actual text! thx
fixing it

mikolak commented on lib/scan_result.dart:
```diff
@@ -14,49 +14,66 @@ abstract class _ScanResultMetadata {
 
   /// An indicator whether the peripheral is connectable (iOS only).
-  bool isConnectable;
+  late final bool isConnectable;
```
Why `late`, as it's initialised in the initialiser list? As it's iOS-only it should be nullable I think.

okocsis replied:
now, that's why it's great that you review this, ofc you are much more aware of the edge cases and behaviours of you own library.
What do you think we move it to an enum instead since it's 3.0 milestone anyways. 
Like `final Connectivity connectivity;` with 3 states:
```Connectivity.noInfo, Connectivity.notConnectable, Connectivity.Connectable```
I'm actually okay with the nullable version as well, but I'm also anxious, people would merge the cases of false and null creating bad logic accidently. It's a common mistake. Thus I generally wouldn't prefer a nullable `bool?` field anywhere.

As the semantic difference of null and false is often unclear and some libs also use them interchangeably, meaning they sloppily emit null instead of false again JS's fault.. the truthy falsy fiasco...

mikolak replied:
Enum sounds OK. I've slapped the milestone 3.0 on this, but I'm actually inclined to release this as 2.4.0, as I don't see any breaking changes, so I'll create a separate issue for the enum.

mikolak replied:
https://github.com/Polidea/FlutterBleLib/issues/588

mikolak commented on lib/scan_result.dart:
```diff
@@ -14,49 +14,66 @@ abstract class _ScanResultMetadata {
 
   /// A list of UUIDs found in the overflow area of the advertisement data (iOS only).
-  List<String> overflowServiceUuids;
+  late final List<String> overflowServiceUuids;
```
Why `late`, as it's initialised in the initialiser list? As it's iOS-only it should be nullable I think.

okocsis replied:
Late is there by mistake, i needed it at first but after refactoring the constructor into an init list it's not needed anymore, thanks to good catch! I'm going to look for all of these unnecessary late qualifiers.
iOS only:
Or it could be understood as default as an empty list on android, therefore not needing to branch the code for iOS/android anymore for the user. 
In other words, does being null carry any useful information for the user here? 
What difference does it make if it's null or it's an empty list of overflowIds semantically? Let's make the caller's life as easy as possible.
On the other hand, if there is a single use-case that the caller would branch on a null differently to an empty list it absolutely must be nullable!

mikolak replied:
Null avoids the issues like `it's always empty` lacking the crucial detail - tested on Android. ;) Though it makes sense to put ease of use over potential unnecessary questions, let's keep it this way.

mikolak commented on lib/scan_result.dart:
```diff
@@ -14,49 +14,66 @@ abstract class _ScanResultMetadata {
+    Map<String, dynamic?> json, 
+    ManagerForPeripheral manager
+  ) {
+    assert(json[_ScanResultMetadata.rssi] is int);
```
Assert only works in debug. 
I don't see the benefit of making it a factory constructor and adding an additional private construct. What's the motivation here?

okocsis replied:
motivations:
-easier to mock a dummy ScanResult object, you just need a @visibleForTesting on a static method that calls this private constructor.
-more importantly, type checker catches _TyepeError-s at the right place
-if something is null out of the dict object it can be concisely handled there in the constructor
-if you want to avoid late qualifiers, the fields must be initialised in the initialiser list, the constructor body -- here factory body -- is too late, so I wanted to avoid to having to us `late` as semantically it's not really late is it?

mikolak replied:
But the original constructor didn't have a body and was public? The only thing changed is (apart from the assert) `overflowServiceUuids ?? <String>[]`.

Doesn't really matter, as this should be changed to abstract class in the future with the implementation moved to src/. Marking as resolved.

mikolak commented on lib/scan_result.dart:
```diff
@@ -14,49 +14,66 @@ abstract class _ScanResultMetadata {
 class AdvertisementData {
   /// The manufacturer data of the peripheral.
-  Uint8List manufacturerData;
+  Uint8List? manufacturerData;
```
Properties of AdvertisementData should be final as well, I guess.

okocsis replied:
right.. and a lots of helper var's on local scope too, I know this `var` thing is just a "bad habit" originated from JS and stuff, the default thing one should type in is `final foo = Bar()` and not `var foo = Bar()` or `Bar foo = Bar()` -- which is also a var -- one should only use var when they mean to reassign it later to some other object.

mikolak commented on lib/src/bridge/characteristics_mixin.dart:
```diff
@@ -169,7 +175,7 @@ mixin CharacteristicsMixin on FlutterBLE {
   }
 
-  Stream<CharacteristicWithValue> monitorCharacteristicForDevice(
+  Stream<CharacteristicWithValueAndTransactionId> monitorCharacteristicForDevice(
```
The motivation here was that the transactionId on the object should be obscured from the consumers, as they expect either Characteristic or CharacteristicWithValue and the transaction id is only used in this file. (this also applies to the rest of the definitions in this file)

okocsis replied:
that's fine, but the test case for some reason expected a CharacteristicWithValueAndTransactionId, and as this is only a static type here and the runtime type was correct ofc, the test still failed, so I played around trying to fix it, and it turned out it was something else, a mocking behaviour changed and that broke the test, so I guess I now can safely change it back and "it SHALL PASS"! ;)

mikolak replied:
I think this is the last thing?

mikolak commented on lib/src/bridge/characteristics_mixin.dart:
```diff
@@ -245,25 +252,31 @@ mixin CharacteristicsMixin on FlutterBLE {
       _parseCharacteristicWithValueWithTransactionIdResponse(
-          Peripheral peripheral, rawJsonValue) {
-    Map<String, dynamic> rootObject = jsonDecode(rawJsonValue);
+          Peripheral peripheral, dynamic? rawJsonValue) {
```
rawJsonValue can be just typed as String, I think? We must have missed type annotation and for reasons unknown Dart didn't complain

okocsis replied:
for rust to complain you need to set static type analysis to strict or something, (I just read this 3day ago).
Anyways, I thought it was dynamic for a reason so I went on and made it dynamic explicitly.
dynamic is really JS mode... :D, drilling a huge hole into the typesystem..

okocsis replied:
I moved the rawJsonValue's String type check to the caller side and made the signature type-safe `String rawJsonValue`.

mikolak commented on lib/src/bridge/scanning_mixin.dart:
```diff
@@ -22,11 +22,13 @@ mixin ScanningMixin on FlutterBLE {
   ) {
-    if (_scanEvents == null) {
-      _prepareScanEventsStream();
+    var scanEvents = _scanEvents;
+    if (scanEvents == null) {
+      scanEvents = _prepareScanEventsStream();
+      _scanEvents = scanEvents;
     }
```
```suggestion
    var scanEvents = _scanEvents ?? _prepareScanEventsStream();
    _scanEvents = scanEvents;
```

mikolak replied:
Or even better: move the second assignment back to `_prepare...` and instead make it return either existing _scanEvents or the newly created, singleton style

okocsis replied:
Okay, so this is annoying me too, it looks bad but that's a trick to make dart actually leverage its nullcheck static capabilities..
Here is the stupid thing in Dart:
if you just check on `if (_scanEvent == null)` then it wouldn't autounwrap _scanEvent after the line it actually gets a statically non-nullable value on line 27 as you recommended, so even still you'd need to unwrap _scanEvent in the local scope like `foo(_scanEvent!)`which is ugly since it's guaranteed to be nonnull.
The trick is if you do the same but with a `local ` and not with an `object field`. 
like so: `var scanEvents = _scanEvents;`
then no manual unwrap is needed the compiler will know that the local var `scanEvents` has a value for sure while it's unsure about its ivar alias `_scanEvents`.

So to sum up you'll see a lot of boilerplate helper code to mitigate this behaviour of Dart, although it's only 2 extra lines each time, and eliminates the need for `force unwrapping` operators, which is semantically noisy and should be avoided!

mikolak replied:
Yeah, I get it, but if _prepareScanEventsStream() always returns non-null and it does the 
```dart
if (_scanEvents == null) { 
   // create a new stream and assign it to _scanEvents
 }
 
 return _scanEvents
 ```
 
you still get that behaviour, but you get rid of the assignment in a function that does not care about it.

okocsis replied:
it will have to be `return _scanEvents!` actually, see below, I'll explain more thoroughly at the point of usage of scanEvenets.

mikolak replied:
I don't mind a helper var in `_prepareScanEventsStream()`, I just don't see the point of having a function return non-nullable if we still need to check for null and reassign in the consumer of said function.

okocsis replied:
Okay gotcha, now _scanEvents is a lazy getter, and _prepareScanEventsStream is gone, this way caller don't need to check anything. I actually would have written it like this myself, just again didn't want to introduce too much change out of the scope of this PR ;)

mikolak commented on test/src/bridge/lib_core_test.dart:
```diff
@@ -9,201 +9,338 @@ import 'package:flutter_ble_lib/src/_internal.dart';
+    );
+    await _emitPlatformError(cancellationErrorJson(mockTransId));
+
+    // await fut;
+    print("anyad");
```
Remove, please

okocsis replied:
:D ups... it was hard fixing tests, so... :D

mikolak reviewed (changes requested):
Got a few questions and a change request, but other than that looks OK!

okocsis committed:
Travis: updated xcode version to 12.2

okocsis committed:
Merge commit '1da3feb70ed4f853145049fe3c5c4ec16080b287' into nullsafety

okocsis committed:
BleError:: default reason text added

okocsis committed:
ScanResult::
added final qualifiers
isConnectable is back to nullable since it's only available on iOS on android it will be null which does not mean false!!

okocsis committed:
CharacteristicsMixin::
_parseCharacteristicWithValueWithTransactionIdResponse()::
signature is now type safe
moved the rawJsonValue String checking to the caller side

okocsis committed:
Test:: removed dummy print

okocsis commented:
@mikolak 
Okay so, it seems the new xcode image did the trick for travis, now phase two passes as well, but still fails on phase three `package-analysis` as follows:
"package-analysis
 It seems that this action doesn't have the required permissions to call the GitHub API with the token you gave. This can occur if this repository is a fork, as in that case GitHub reduces the GITHUB_TOKEN's permissions for security reasons. Consequently, no report will be made on GitHub. Check this issue for more information: https://github.com/axel-op/dart-package-analyzer/issues/2"
 
 I don't know how to fix this one mate.

mikolak commented:
Yeah, it's because it's a PR from a fork instead of a MR from this repo - don't mind.

okocsis commented on lib/src/bridge/scanning_mixin.dart:
```diff
@@ -40,8 +42,8 @@ mixin ScanningMixin on FlutterBLE {
     );
 
     streamController
-        .addStream(_scanEvents, cancelOnError: true)
```
@mikolak the problem lies here. If I leave it _scanEvents then the compiler is just not clever enough to know that _scanEvents is actually autoUnwrappable, and I'll have to write: `.addStream(_scanEvents!, cancelOnError: true)`
It's not that force unwrap operator is ugly, it's discouraged to use for a reason:
- Imagine, if the code above changes any time in the future, changing the behaviour of the code in way that `_scanEvents` is no longer guaranteed to be `nonnull`.
Nothing is gonna warn you about the `!` force unwrap now being semantically incorrect, and crashing the software on edge cases.
- Whereas if you maintain a helper local var instead /more verbosely I know/, then this behaviour change on 43 will break the compilation of the code since the compiler will see that `scanEvents` local var can no longer be proven `nonnull` and thus it will be `nullable` and `addStream()` don't accept no nullable value by its signature, so it will not compile, which is a good thing. It can safeguard you from future mistakes even, warning you to add an extra null check or something.

So yeah, it may seem to be an unnecessary complication with the extra var and all, but it really for semanticly correct nullsafe coding. 
Ofc, I'm also happy with anything that achieves the same, correct me if I'm wrong but I don't see you suggested change achieving the same safety concern.
Or b.) no fucks given and I should just make it less verbose sacrificing nullsafety.
Up to you ;)

okocsis committed:
ScanningMixin:: _prepareScanEventsStream() renamed to just _scanEvents
also _resetScanEvents introduced to null the underlying scream out

mikolak commented on lib/src/bridge/scanning_mixin.dart:
```diff
@@ -1,19 +1,29 @@
-        .receiveBroadcastStream()
-        .handleError(
-          (errorJson) => throw BleError.fromJson(jsonDecode(errorJson.details)),
+  Stream<ScanResult>? __scanEvents;
```
How about `_activeScanEvents` or `_currentScanStream`? I don't like having a variable discerned only by having one more dash in front - too easy to fool the eye, especially with IDE suggestion.

okocsis replied:
the goal was to make it super-duper private, and to denote that it wasn't meant to be used from the class scope only from the scope of the actual interface being:
`get _scanEvents` getter and `_resetScanEvents()` to indirectly interface that state variable in the class.
`_activeScanEvents` works too ofc. ;)

mikolak commented on lib/src/bridge/scanning_mixin.dart:
```diff
@@ -1,19 +1,29 @@
+      scanEvents = 
+        const EventChannel(
+          ChannelName.scanningEvents
+        ).receiveBroadcastStream(
+        ).handleError(
```
Has the formatter been run here? It looks weird.

okocsis replied:
it's a style, it's different to your style, to me it's easier to overlook function chaining like this because `).foo()`easier to catch for the eye then simply `.foo()` wich as you start to read it can also be an enum case with a leading dot /not in dart ofc/.. I don't mind changing it to just leading dot style, but it already wasn't consistent anyways.

I do not like autoformatters (or i haven't found one that actually works well)

mikolak replied:
pub.dev cuts points for code not formatted using the dartfmt. Like it or not, it's mandatory. 🙂

okocsis committed:
ScanningMixin:: small refinement on naming stuff

okocsis commented on lib/src/bridge/characteristics_mixin.dart:
```diff
@@ -66,11 +77,16 @@ mixin CharacteristicsMixin on FlutterBLE {
+            );
+          });
 
   Future<void> writeCharacteristicForIdentifier(
```
@mikolak 
OFF-TOPIC:
it looks like the response is swallowed here and it returns void instead of `Characteristic? response`
is this intentional?

Same in other write methods where the response is optionally available, all those should return `Future<CharacteristicsWithValue?>` as opposed to `Future<Characteristics>` right?

Or am I overlooking something? 🤔

okocsis replied:
the last commit is to mitigate this. 
Pls recommend a better exception to throw on null value erroneously coming from native.
Thanks!

okocsis committed:
Characteristic:: refactor on Futures to always complete with something even if withResponse is false
additional type safety on invokeMethod calls

mikolak commented on lib/src/bridge/bluetooth_state_mixin.dart:
```diff
@@ -37,7 +39,7 @@ mixin BluetoothStateMixin on FlutterBLE {
   }
 
-  BluetoothState _mapToBluetoothState(dynamic rawValue) {
+  BluetoothState _mapToBluetoothState(String? rawValue) {
```
Unless Dart complains, this should be typed as non-nullable. If for any reason a null came from the native side, then it's an error and should not be swept under the rug.

mikolak commented on lib/service.dart:
```diff
@@ -37,7 +37,7 @@ class Service extends InternalService {
   /// [Characteristic.isWritableWithoutResponse] is `true` and
   /// [withResponse] is specified accordingly can be written to.
-  Future<Characteristic> writeCharacteristic(
+  Future<Characteristic?> writeCharacteristic(
```
This is not nullable, it's either characteristic or error

mikolak commented on lib/src/_managers_for_classes.dart:
```diff
@@ -52,7 +52,7 @@ abstract class ManagerForPeripheral {
   );
 
-  Future<Characteristic> writeCharacteristicForDevice(
+  Future<Characteristic?> writeCharacteristicForDevice(
```
This is not nullable, it's either characteristic or error

mikolak commented on lib/src/_managers_for_classes.dart:
```diff
@@ -101,7 +101,7 @@ abstract class ManagerForService {
   );
 
-  Future<Characteristic> writeCharacteristicForService(
+  Future<Characteristic?> writeCharacteristicForService(
```
This is not nullable, it's either characteristic or error

mikolak commented on lib/src/_managers_for_classes.dart:
```diff
@@ -145,7 +145,7 @@ abstract class ManagerForCharacteristic {
   );
 
-  Future<void> writeCharacteristicForIdentifier(
+  Future<Characteristic?> writeCharacteristicForIdentifier(
```
This is not nullable, it's either characteristic or error

mikolak commented on lib/peripheral.dart:
```diff
@@ -143,7 +143,7 @@ class Peripheral {
   ///
   /// Will result in error if discovery was not done during this connection.
-  Future<Characteristic> writeCharacteristic(
+  Future<Characteristic?> writeCharacteristic(
```
This is not nullable, it's either characteristic or error

mikolak commented on lib/characteristic.dart:
```diff
@@ -69,7 +69,7 @@ class Characteristic extends InternalCharacteristic {
   /// [isWritableWithoutResponse] is `true` and argument [withResponse] is
   /// set accordingly.
-  Future<void> write(
+  Future<Characteristic?> write(
```
Void here was intentional. There's a bug with not returning response from write with response, but it has to be fixed on multiple layers and even then this would be typed as `Uint8List`.

mikolak commented on lib/src/bridge/bluetooth_state_mixin.dart:
```diff
@@ -1,9 +1,11 @@
+  final Stream<String> _adapterStateChanges =
+    const EventChannel(ChannelName.adapterStateChanges)
+      .receiveBroadcastStream()
+      .where((event) => event is String)
```
There shouldn't be any filtering on this stream. If for any reason an unexpected value came, it should be raised as an error.

mikolak commented on lib/src/bridge/characteristics_mixin.dart:
```diff
@@ -11,23 +11,23 @@ mixin CharacteristicsMixin on FlutterBLE {
-            String rawJsonValue = "";
-            if (rawValue is String) {
-              rawJsonValue = rawValue;
+            if (rawValue == null) {
```
If characteristic didn't come, then an exception should be raised. Returning some invalid default value only obscures the situation.

okocsis replied:
That's fine. I understand, you see the correct semantics for this domain much better. It's just that, there were some -- and correct me if I'm wrong where a function returning a future, could become into a hanging state and never actually resolve, only on timeout if the caller sets one. Or is it always guaranteed from the native side that 'some' `.then()` (or `error()`) will always be executed?

The reason I don't get this is that there is this `withResponse` flag user can pass in and in the case of that being false, how will the returned future behave??
To me, it looks like it will never resolve without a(n unnecessary) timeout in the case of `withResponse = false`, and even then it wouldn't return a characteristic in the future raising a _TypeError in new typesafer dart runtime.

and on `withResponse = true` it behaves as expected ofc

This is why I thought it would make sense to make it return a nullable resolving future, that is guaranteed to always resolve, making the caller code much cleaner.

To sum up: 
- I will revert the commit, but pls consider these concerns, as these are changes better be done on a major version bump, as it introduces minor API amendments, potentially breaking adopters code also.
- What does Future<Characteristic> resolve to exactly on `withResponse = false`?

thanks

mikolak replied:
Response is a byte array from the server (peripheral). If the central (client) does not await for the response, the operation is finished immediately after sending the write request.

Native plugin always returns either characteristic or error. The whole characteristic is returned regardless of the response for convenience - this way you can write to a characteristic from peripheral, save the reference to the characteristic and continue operations on it.

mikolak commented on lib/src/bridge/characteristics_mixin.dart:
```diff
@@ -47,16 +47,16 @@ mixin CharacteristicsMixin on FlutterBLE {
-            if (rawValue is String) {
-              rawJsonValue = rawValue;
+            if (rawValue == null) {
+              throw Exception("rawValue cannot be null");
```
This exception doesn't give any value to the consumer of the API.

okocsis replied:
yes, I know. I wanted to discuss what it should throw, but now it's postponed as I've reverted the commit. ;)

mikolak reviewed (changes requested):
I appreciate your work, but please, revert the last commit. Native client has to either return characteristic or error and I'd rather have a null pointer exception thrown than some custom exception that does not give the API consumer any insight or returning an invalid default value.

There's a bug where the response is not properly returned on Android, but it's a multilayer issue that originates in Android's MBA, any changes to interface will be done along with fixing it.

Changes to the mixins are unrelated to this PR and should be in a separate PR.

Summing up: thanks for all the work on null safety! ❤️  Please, run the formatter in scanning_mixin.dart and revert commit `86dafd6` (`Characteristic:: refactor on Futures to always complete...`) and I'll merge it! 🙂

okocsis committed:
Revert "Characteristic:: refactor on Futures to always complete with something even if withResponse is false"

This reverts commit 86dafd6d0dc4596b03150aeade1a09dfb29596ac.

okocsis committed:
Merge commit 'aaa2009274a0b6c48a0eeae7565143f87b89ee2c' into nullsafety

# Conflicts:
#	lib/src/bridge/lib_core.dart

mikolak commented on lib/src/bridge/characteristics_mixin.dart:
```diff
@@ -20,10 +20,16 @@ mixin CharacteristicsMixin on FlutterBLE {
-                  .value);
+          .then((rawValue) {
+            String rawJsonValue = "";
+            if (rawValue is String) {
```
This can actually be get rid of by changing to `.invokeMethod<String>(...)`

okocsis replied:
yeah, so I figured.  As later I went and explored invokeMethod nullsafe amendments..
I did those on the latest commit that ended up being reverted. :)
we can refactor the whole code with this making it much more type correct! ;)

mikolak commented on lib/src/bridge/characteristics_mixin.dart:
```diff
@@ -43,11 +49,16 @@ mixin CharacteristicsMixin on FlutterBLE {
-                    peripheral, rawJsonValue),
-          );
+          .then((rawValue) {
+            String rawJsonValue = "";
+            if (rawValue is String) {
+              rawJsonValue = rawValue;
+            }
```
This can actually be get rid of by changing to `.invokeMethod<String>(...)`
And yes, it's guaranteed to be string at this point in time.

okocsis replied:
Yes, now I know, I didn't know it then when I wrote this... :/ would have been much more simpler to fixup the bit too loose typing in the code :). But that's how we learn right?

okocsis replied:
> it's guaranteed to be string at this point in time.

As a matter of fact it will have the correct type inferred even of `rawValue` no need to statically denote the type in `then()`-s param list like so:
 - good: `then(rawValue)` inferred to be string cause of invoke's generic type notation
 - bad: `then(String rawValue)` unnecessary and can even break when `invokeMetod<dynamic>`

mikolak commented on lib/src/bridge/characteristics_mixin.dart:
```diff
@@ -66,11 +77,16 @@ mixin CharacteristicsMixin on FlutterBLE {
-                    peripheral, rawJsonValue),
-          );
+          .then((rawValue) {
+            String rawJsonValue = "";
+            if (rawValue is String) {
+              rawJsonValue = rawValue;
+            }
```
This can actually be get rid of by changing to `.invokeMethod<String>(...)`
And yes, it's guaranteed to be string at this point in time.

mikolak commented on lib/src/bridge/characteristics_mixin.dart:
```diff
@@ -230,39 +252,47 @@ mixin CharacteristicsMixin on FlutterBLE {
+    Stream<CharacteristicWithValueAndTransactionId> stream =
+      _characteristicsMonitoringEvents
+        .map((rawValue) {
+          String rawJsonValue = "";
+          if (rawValue is String) {
+            rawJsonValue = rawValue;
+          }
```
This can actually be get rid of by changing to `.invokeMethod<String>(...)`
And yes, it's guaranteed to be string at this point in time.

mikolak commented on lib/src/internal_ble_manager.dart:
```diff
@@ -185,20 +184,20 @@ class InternalBleManager
   Future<List<Peripheral>> knownPeripherals(
       List<String> peripheralIdentifiers) {
-    return _bleLib.knownDevices(peripheralIdentifiers ?? []);
+    return _bleLib.knownDevices(peripheralIdentifiers);
```
Might need confirmation

okocsis replied:
the confirmation lays underneath ofc not here, but if one can pass in a null then they might as well just pass in an empty list. 
Null on input is not something desirable unless it actually carries a different semantic meaning.
so basically it's just guaranteed by the type system and not this extra safety code.

mikolak commented on lib/src/internal_ble_manager.dart:
```diff
@@ -185,20 +184,20 @@ class InternalBleManager
   @override
   Future<List<Peripheral>> connectedPeripherals(List<String> serviceUuids) {
-    return _bleLib.connectedDevices(serviceUuids ?? []);
+    return _bleLib.connectedDevices(serviceUuids);
```
Might need confirmation

okocsis replied:
same here List<String> serviceUuids doesn't allow a null value

mikolak reviewed (approved):
A few minors, but I'll take care of those myself. Thank you very much for the PR!
